File: src/core/cloudflare-context.ts
import { CloudflareActions } from "./cloudflare-actions.js";
import type { ConfigStore } from "./config-store.js";
import { CLOUDFLARE_AUTH, CLOUDFLARE_PROVIDER_ID } from "./cloudflare-auth.js";
import { CloudflareManager, type CloudflareProject } from "./cloudflare-manager.js";
import {
  CLOUDFLARE_HOOKS,
  CLOUDFLARE_MANIFEST,
  createCloudflareDeployActions,
  createCloudflareDeployProvider,
  toProviderProject,
} from "./cloudflare-provider.js";
import { resolveProviderCredential, type ProviderCredential } from "./providers/credentials.js";
import type { ProviderProject } from "./providers/deploy-provider.js";
import {
  adoptSoleScope,
  resolveProviderProject,
} from "./providers/project-resolution.js";
import type { ProviderContext, RegisteredDeployProvider } from "./providers/registry.js";
import type { NoMoreIdeConfig } from "./types.js";

/**
 * Cloudflare's registry entry — how a connected client is built for a working
 * directory, and the write-capable half resolved separately from it.
 *
 * The registry imports this; this must not import the registry, or the two
 * would cycle.
 *
 * Shorter than Vercel's counterpart in one place and longer in another, which
 * is the point of doing provider #2 before publishing anything:
 *
 * - **No token persistence.** Cloudflare has no browser sign-in, so nothing
 *   here ever rotates a refresh token — the `onTokensRefreshed` plumbing Vercel
 *   needs is simply absent.
 * - **The account is resolved before the project, not after.** Every Pages path
 *   contains an account id, so project resolution cannot even begin until the
 *   scope is known.
 */

/**
 * A connected Cloudflare client for the given working directory.
 *
 * Throws when Cloudflare isn't connected. A missing *account* is not an error
 * either: `account()` and `listScopes()` answer without one, which is what lets
 * the dashboard show the account switcher rather than a dead end.
 */
async function cloudflareContext(
  configStore: ConfigStore,
  gitCwd: string,
): Promise<ProviderContext> {
  const config = await configStore.load();
  const credential = await resolveProviderCredential(CLOUDFLARE_AUTH, config, process.env);
  const accountId = credential.scopeId ?? (await adoptDefaultAccount(configStore, credential));
  const manager = new CloudflareManager(credential.token, accountId);
  const project = await resolveProject(config, manager, gitCwd);
  return {
    // The project is resolved first and handed to the provider: Cloudflare has
    // no global deployment endpoint, so the deployment reads need it.
    provider: createCloudflareDeployProvider(manager, project?.name),
    credential: { ...credential, scopeId: accountId },
    project,
  };
}

/**
 * The account to use when the user has never chosen one.
 *
 * `cliSelectsScope: false` because Wrangler has no account switch to inherit —
 * so unlike Vercel, a `cli` credential does adopt the sole account, or a
 * `wrangler login` with one account could never read a single project. It is
 * still not written to config, so `CLOUDFLARE_ACCOUNT_ID` keeps winning.
 */
function adoptDefaultAccount(
  configStore: ConfigStore,
  credential: ProviderCredential,
): Promise<string | undefined> {
  return adoptSoleScope({
    source: credential.source,
    cliSelectsScope: false,
    listScopes: () =>
      new CloudflareManager(credential.token).listAccounts().then((accounts) =>
        // Cloudflare accounts have no slug; the id addresses one.
        accounts.map((account) => ({ id: account.id, slug: account.id })),
      ),
    persist: (scope) =>
      configStore.setConnectionScope(CLOUDFLARE_PROVIDER_ID, scope).then(() => undefined),
  });
}

/** Write-capable counterpart, resolved the same way but returned separately. */
async function cloudflareActions(configStore: ConfigStore) {
  const config = await configStore.load();
  const credential = await resolveProviderCredential(CLOUDFLARE_AUTH, config, process.env);
  const accountId = credential.scopeId ?? (await adoptDefaultAccount(configStore, credential));
  if (!accountId) {
    throw new Error("Choose a Cloudflare account before changing one of its projects.");
  }
  return createCloudflareDeployActions(
    new CloudflareActions({ token: credential.token }, accountId),
  );
}

/**
 * Which Cloudflare project this repository deploys — the shared three-tier
 * ladder, with only two of its three tiers in play: Wrangler writes no link
 * file, so an unpinned repo is matched on its git remote.
 */
async function resolveProject(
  config: NoMoreIdeConfig,
  manager: CloudflareManager,
  gitCwd: string,
): Promise<ProviderProject | undefined> {
  const project = await resolveProviderProject<CloudflareProject>({
    providerId: CLOUDFLARE_PROVIDER_ID,
    hooks: CLOUDFLARE_HOOKS,
    config,
    gitCwd,
    getProject: (id) => manager.getProject(id),
    findByRepoUrl: (repoUrl) => manager.findProjectByRepo(repoUrl),
  });
  return project ? toProviderProject(project) : undefined;
}

export const cloudflareDeployProvider: RegisteredDeployProvider = {
  manifest: CLOUDFLARE_MANIFEST,
  auth: CLOUDFLARE_AUTH,
  hooks: CLOUDFLARE_HOOKS,
  context: cloudflareContext,
  actions: cloudflareActions,
};

File: crates/nomoreide-core/src/cloudflare_context.rs
//! Resolves "which Cloudflare account does this token mean" — the Rust
//! counterpart of `src/core/cloudflare-context.ts`.

use serde_json::Value;

use crate::cloudflare_actions::CloudflareActions;
use crate::cloudflare_auth::{resolve, ResolvedCredential, CLOUDFLARE_PROVIDER_ID};
use crate::cloudflare_manager::CloudflareManager;
use crate::config::{Config, ConfigStore};

/// A connected Cloudflare client and the credential behind it.
pub async fn require_client(
    store: &ConfigStore,
    config: &Config,
) -> Result<(CloudflareManager, ResolvedCredential), String> {
    let credential = resolve(store, config).await?;
    let account_id = match credential.account_id.clone() {
        Some(account_id) => Some(account_id),
        None => adopt_sole_account(store, &credential).await,
    };
    let manager = CloudflareManager::new(credential.token.clone(), account_id.clone());
    Ok((
        manager,
        ResolvedCredential {
            account_id,
            ..credential
        },
    ))
}

/// Write-capable counterpart, resolved the same way but returned separately so
/// the read/write split is visible at the call site.
///
/// Unlike the read client, this one *requires* an account: every Pages write is
/// a `PATCH` of `/accounts/<id>/…`, and there is no useful thing to do without
/// one.
pub async fn require_actions(
    store: &ConfigStore,
    config: &Config,
) -> Result<CloudflareActions, String> {
    let credential = resolve(store, config).await?;
    let account_id = match credential.account_id.clone() {
        Some(account_id) => account_id,
        None => adopt_sole_account(store, &credential)
            .await
            .ok_or("Choose a Cloudflare account before changing one of its projects.")?,
    };
    Ok(CloudflareActions::new(credential.token, account_id))
}

/// The account to use when the user has never chosen one.
///
/// Every Pages path is `/accounts/<id>/…`, so a client without one can do
/// nothing at all — which is why the sole account is adopted rather than left
/// for the scope switcher to ask about. With several there is no unambiguous
/// answer, so none is chosen.
///
/// **A `cli` credential adopts but does not persist.** Wrangler has no account
/// switch of its own; the account comes from `CLOUDFLARE_ACCOUNT_ID` in the
/// environment, and a saved scope would outrank that variable from then on.
///
/// Failures are swallowed: an unscoped client still reports the "choose an
/// account" message, and the caller's own request should be what surfaces an
/// auth problem.
async fn adopt_sole_account(
    store: &ConfigStore,
    credential: &ResolvedCredential,
) -> Option<String> {
    let accounts = CloudflareManager::new(credential.token.clone(), None)
        .list_accounts()
        .await
        .ok()?;
    if accounts.len() != 1 {
        return None;
    }
    let account = accounts.first()?;
    let id = account.get("id").and_then(Value::as_str)?.to_string();
    if credential.source != "cli" {
        let name = account
            .get("name")
            .and_then(Value::as_str)
            .map(str::to_string);
        store
            .set_connection_scope(CLOUDFLARE_PROVIDER_ID, Some(id.clone()), name)
            .await
            .ok()?;
    }
    Some(id)
}
